ecs-cli/modules/cli/cluster/userdata/user_data_test.go

```diff
@@ -0,0 +1,204 @@
+
+var extraUserDataShellScript = `#!/bin/bash
+echo "Quickly, the honeymoon bliss wears off" >> $HOME/chapter4
+echo "The once happy couple, now fights over small things, like who gets to chew which toy" >> $HOME/chapter4
+echo "Pudding drags her bed to away from Tum Tum's, and hides a pile of toys underneath it" >> $HOME/chapter4`
+
+var extraUserDataCloudConfig = `#cloud-config
+
+# Install additional packages on first boot
+packages:
+ - pwgen
+ - pastebinit
+ - [libpython2.7, 2.7.3-0ubuntu3.1]
+ - gpg`
+
+var expectedMimeMultipart = `Content-Type: multipart/mixed; boundary="========multipart-boundary=="
+MIME-Version: 1.0
+
+--========multipart-boundary==
+Content-Type: text/text/plain; charset="utf-8"
+Mime-Version: 1.0
+
+#!/bin/bash
+echo "Clyde spends the next year lonely and confused" >> $HOME/chapter3
+echo "He knows Pudding wasn't right for him, but he doesn't know how to move on" >> $HOME/chapter3
+echo "Meanwhile, Pudding and Tum Tum have the honeymoon of a lifetime" >> $HOME/chapter3
+echo "They spend 3 months travelling the world, visiting DogseyLand, the Temple of Doge in Japan, and much more" >> $HOME/chapter3
+echo "Tum Tum experiments with 10 different haircuts/stylings" >> $HOME/chapter3
+
+--========multipart-boundary==
+Content-Type: text/text/x-shellscript; charset="utf-8"
+Mime-Version: 1.0
+
+
+#!/bin/bash
+aws s3 cp s3://my-scripts/setupenv $HOME/setupenv
+source $HOME/setupenv
+
+
+--========multipart-boundary==
+Content-Type: text/text/plain; charset="utf-8"
+Mime-Version: 1.0
+
+#!/bin/bash
+echo "Quickly, the honeymoon bliss wears off" >> $HOME/chapter4
+echo "The once happy couple, now fights over small things, like who gets to chew which toy" >> $HOME/chapter4
+echo "Pudding drags her bed to away from Tum Tum's, and hides a pile of toys underneath it" >> $HOME/chapter4
+--========multipart-boundary==
+Content-Type: text/text/plain; charset="utf-8"
+Mime-Version: 1.0
+
+#cloud-config
+
+# Install additional packages on first boot
+packages:
+ - pwgen
+ - pastebinit
+ - [libpython2.7, 2.7.3-0ubuntu3.1]
+ - gpg
+--========multipart-boundary==
+Content-Type: text/text/x-shellscript; charset="utf-8"
+Mime-Version: 1.0
+
+
+#!/bin/bash
+echo ECS_CLUSTER=cluster >> /etc/ecs/ecs.config
+
+--========multipart-boundary==--
+`
+
+func TestBuildUserDataWithExtraData(t *testing.T) {
+
+	multipartFilePath := writeTempFile(t, "existingMultipartArchive", existingMultipartArchive)
+	defer os.Remove(multipartFilePath)
+
+	shellScriptFilePath := writeTempFile(t, "extraUserDataShellScript", extraUserDataShellScript)
+	defer os.Remove(shellScriptFilePath)
+
+	cloudConfigFilePath := writeTempFile(t, "extraUserDataCloudConfig", extraUserDataCloudConfig)
+	defer os.Remove(cloudConfigFilePath)
+
+	buf := new(bytes.Buffer)
+	writer := multipart.NewWriter(buf)
+	// set the boundary between parts so that output is deterministic
+	writer.SetBoundary(testBoundary)
+	builder := newBuilderInTest(buf, writer)
```

Comment:
Can you clarify why we are mocking out the builder and only testing methods on the mock? Not quite sure what the point of the test is if it's not running the real code. Looking at `NewBuilder`, I'm not totally sure I understand why this needs to be mocked.

Reply:
`newBuilderInTest` doesn't return a mock, it returns the actual builder object, so we're testing the real code. This is needed so that we can set the multipart boundary on the writer that the builder uses (otherwise the boundary would be random, and this would make the output non-deterministic, and we couldn't compare it with an expected value in the test).